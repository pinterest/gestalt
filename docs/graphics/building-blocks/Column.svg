<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><path fill="#4A4A4A" fill-rule="evenodd" d="M79 0a1 1 0 0 1 1 1v4.045a1 1 0 1 1-2 0V1a1 1 0 0 1 1-1Zm0 12.136a1 1 0 0 1 1 1v8.091a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1ZM79 44.5a1 1 0 0 1 1 1v8.09a1 1 0 1 1-2 0V45.5a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.181a1 1 0 0 1 1 1v8.091a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1V179a1 1 0 0 1-2 0v-4.045a1 1 0 0 1 1-1ZM157 0a1 1 0 0 1 1 1v4.045a1 1 0 0 1-2 0V1a1 1 0 0 1 1-1Zm0 12.136a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0V45.5a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.181a1 1 0 0 1 1 1v8.091a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1V179a1 1 0 0 1-2 0v-4.045a1 1 0 0 1 1-1ZM235 0a1 1 0 0 1 1 1v4.045a1 1 0 0 1-2 0V1a1 1 0 0 1 1-1Zm0 12.136a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0V45.5a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.09a1 1 0 0 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.181a1 1 0 0 1 1 1v8.091a1 1 0 1 1-2 0v-8.09a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1v8.091a1 1 0 0 1-2 0v-8.091a1 1 0 0 1 1-1Zm0 16.182a1 1 0 0 1 1 1V179a1 1 0 0 1-2 0v-4.045a1 1 0 0 1 1-1Z" clip-rule="evenodd"/></svg>