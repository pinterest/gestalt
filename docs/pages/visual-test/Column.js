// @flow strict
import { type Node as ReactNode } from 'react';
import { Box, Column, Flex, Text } from 'gestalt';

export default function Snapshot(): ReactNode {
  return (
    <Box color="default" display="inlineBlock" padding={1} width={400}>
      <Flex direction="column">
        <Column span={4}>
          <Box color="infoWeak" margin={1} padding={4}>
            <Text align="center">Column 1</Text>
          </Box>
        </Column>
        <Column span={4}>
          <Box color="infoWeak" margin={1} padding={4}>
            <Text align="center">Column 2</Text>
          </Box>
        </Column>
        <Column span={4}>
          <Box color="infoWeak" margin={1} padding={4}>
            <Text align="center">Column 3</Text>
          </Box>
        </Column>
      </Flex>
      <Flex>
        <Column span={4}>
          <Box color="successWeak" margin={1} padding={4}>
            <Text align="center">Column 1</Text>
          </Box>
        </Column>
        <Column span={4}>
          <Box color="successWeak" margin={1} padding={4}>
            <Text align="center">Column 2</Text>
          </Box>
        </Column>
        <Column span={4}>
          <Box color="successWeak" margin={1} padding={4}>
            <Text align="center">Column 3</Text>
          </Box>
        </Column>
      </Flex>
    </Box>
  );
}
