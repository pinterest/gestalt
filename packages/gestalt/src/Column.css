:root {
  --ncols: 12;
}

.col {
  vertical-align: top;
}

.xsCol1 {
  width: calc(1 / var(--ncols) * 100%);
}

.xsCol2 {
  width: calc(2 / var(--ncols) * 100%);
}

.xsCol3 {
  width: calc(3 / var(--ncols) * 100%);
}

.xsCol4 {
  width: calc(4 / var(--ncols) * 100%);
}

.xsCol5 {
  width: calc(5 / var(--ncols) * 100%);
}

.xsCol6 {
  width: calc(6 / var(--ncols) * 100%);
}

.xsCol7 {
  width: calc(7 / var(--ncols) * 100%);
}

.xsCol8 {
  width: calc(8 / var(--ncols) * 100%);
}

.xsCol9 {
  width: calc(9 / var(--ncols) * 100%);
}

.xsCol10 {
  width: calc(10 / var(--ncols) * 100%);
}

.xsCol11 {
  width: calc(11 / var(--ncols) * 100%);
}

.xsCol12 {
  width: calc(12 / var(--ncols) * 100%);
}

@media (--sm) {
  .smCol1 {
    width: calc(1 / var(--ncols) * 100%);
  }

  .smCol2 {
    width: calc(2 / var(--ncols) * 100%);
  }

  .smCol3 {
    width: calc(3 / var(--ncols) * 100%);
  }

  .smCol4 {
    width: calc(4 / var(--ncols) * 100%);
  }

  .smCol5 {
    width: calc(5 / var(--ncols) * 100%);
  }

  .smCol6 {
    width: calc(6 / var(--ncols) * 100%);
  }

  .smCol7 {
    width: calc(7 / var(--ncols) * 100%);
  }

  .smCol8 {
    width: calc(8 / var(--ncols) * 100%);
  }

  .smCol9 {
    width: calc(9 / var(--ncols) * 100%);
  }

  .smCol10 {
    width: calc(10 / var(--ncols) * 100%);
  }

  .smCol11 {
    width: calc(11 / var(--ncols) * 100%);
  }

  .smCol12 {
    width: calc(12 / var(--ncols) * 100%);
  }
}

@media (--md) {
  .mdCol1 {
    width: calc(1 / var(--ncols) * 100%);
  }

  .mdCol2 {
    width: calc(2 / var(--ncols) * 100%);
  }

  .mdCol3 {
    width: calc(3 / var(--ncols) * 100%);
  }

  .mdCol4 {
    width: calc(4 / var(--ncols) * 100%);
  }

  .mdCol5 {
    width: calc(5 / var(--ncols) * 100%);
  }

  .mdCol6 {
    width: calc(6 / var(--ncols) * 100%);
  }

  .mdCol7 {
    width: calc(7 / var(--ncols) * 100%);
  }

  .mdCol8 {
    width: calc(8 / var(--ncols) * 100%);
  }

  .mdCol9 {
    width: calc(9 / var(--ncols) * 100%);
  }

  .mdCol10 {
    width: calc(10 / var(--ncols) * 100%);
  }

  .mdCol11 {
    width: calc(11 / var(--ncols) * 100%);
  }

  .mdCol12 {
    width: calc(12 / var(--ncols) * 100%);
  }
}

@media (--lg) {
  .lgCol1 {
    width: calc(1 / var(--ncols) * 100%);
  }

  .lgCol2 {
    width: calc(2 / var(--ncols) * 100%);
  }

  .lgCol3 {
    width: calc(3 / var(--ncols) * 100%);
  }

  .lgCol4 {
    width: calc(4 / var(--ncols) * 100%);
  }

  .lgCol5 {
    width: calc(5 / var(--ncols) * 100%);
  }

  .lgCol6 {
    width: calc(6 / var(--ncols) * 100%);
  }

  .lgCol7 {
    width: calc(7 / var(--ncols) * 100%);
  }

  .lgCol8 {
    width: calc(8 / var(--ncols) * 100%);
  }

  .lgCol9 {
    width: calc(9 / var(--ncols) * 100%);
  }

  .lgCol10 {
    width: calc(10 / var(--ncols) * 100%);
  }

  .lgCol11 {
    width: calc(11 / var(--ncols) * 100%);
  }

  .lgCol12 {
    width: calc(12 / var(--ncols) * 100%);
  }
}
