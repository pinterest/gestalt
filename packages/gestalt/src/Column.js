/* @flow */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './ColumnColumn.css';

type DeprecatedColumns = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
type Columns = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

type ColumnProps =
  | {|
      children?: React.Node,
      span: Columns,
      smSpan?: Columns,
      mdSpan?: Columns,
      lgSpan?: Columns,
    |}
  | {|
      children?: React.Node,
      xs?: DeprecatedColumns,
      sm?: DeprecatedColumns,
      md?: DeprecatedColumns,
      lg?: DeprecatedColumns,
    |};

export default function Column(props: ColumnProps) {
  const { children } = props;
  const cs = classnames(
    (props.xs !== undefined ||
      props.sm !== undefined ||
      props.md !== undefined ||
      props.lg !== undefined) &&
      styles.deprecatedColumn,
    props.xs && styles[`xsCol${props.xs}`],
    props.sm && styles[`smCol${props.sm}`],
    props.md && styles[`mdCol${props.md}`],
    props.lg && styles[`lgCol${props.lg}`],
    props.span != null && styles[`xsCol${props.span}`],
    props.smSpan != null && styles[`smCol${props.smSpan}`],
    props.mdSpan != null && styles[`mdCol${props.mdSpan}`],
    props.lgSpan != null && styles[`lgCol${props.lgSpan}`]
  );
  return <div className={cs}>{children}</div>;
}

Column.propTypes = {
  children: PropTypes.node,
  xs: PropTypes.number,
  sm: PropTypes.number,
  md: PropTypes.number,
  lg: PropTypes.number,
  span: PropTypes.number,
  smSpan: PropTypes.number,
  mdSpan: PropTypes.number,
  lgSpan: PropTypes.number,
};
