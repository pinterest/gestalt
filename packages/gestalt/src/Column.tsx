import { ReactNode } from 'react';
import classnames from 'classnames';
import styles from './Column.css';

type Columns = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

type ColumnProps = {
  /**
   * The content to be laid out.
   */
  children?: ReactNode;
  /**
   * The number of units in a 12-unit width that this element will occupy.
   *
   * Also available in responsive sizes: `smSpan`, `mdSpan`, `lgSpan`
   */
  span: Columns;
  /**
   * The number of units in a 12-unit width that this element will occupy in sm and larger viewports.
   */
  smSpan?: Columns;
  /**
   * The number of units in a 12-unit width that this element will occupy in md and larger viewports.
   */
  mdSpan?: Columns;
  /**
   * The number of units in a 12-unit width that this element will occupy in lg and larger viewports.
   */
  lgSpan?: Columns;
};

/**
 * Use [Column](https://gestalt.pinterest.systems/web/column) to implement a 12-column system.
 *
 * ![Column light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Column.spec.mjs-snapshots/Column-chromium-darwin.png)
 *
 */
export default function Column(props: ColumnProps) {
  const { children } = props;
  const cs = classnames(
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore cannot infer type with dynamic property name
    props.span != null && styles[`xsCol${props.span}`],
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore cannot infer type with dynamic property name
    props.smSpan != null && styles[`smCol${props.smSpan}`],
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore cannot infer type with dynamic property name
    props.mdSpan != null && styles[`mdCol${props.mdSpan}`],
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore cannot infer type with dynamic property name
    props.lgSpan != null && styles[`lgCol${props.lgSpan}`],
  );
  return <div className={cs}>{children}</div>;
}

Column.displayName = 'Column';
