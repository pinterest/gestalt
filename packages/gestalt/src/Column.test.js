// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import Column from './Column.js';

test('Column deprecated span', () => {
  const component = renderer.create(<Column xs={1}>Hello world</Column>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Column span', () => {
  const component = renderer.create(<Column span={1}>Hello world</Column>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
