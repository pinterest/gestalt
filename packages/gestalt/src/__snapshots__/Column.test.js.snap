// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column span 1`] = `
<div
  className="xsCol1"
>
  Hello world
</div>
`;
